Sources/performance-worker/WorkerService.swift

```diff
@@ -114,98 +68,176 @@ final class WorkerService: Grpc_Testing_WorkerService.ServiceProtocol, Sendable
-      switch self.role {
-      case .server(_):
-        throw RPCError(code: .alreadyExists, message: "A server has already been set up.")
+    enum OnStartedServer {
+      case runServer
+      case invalidState(RPCError)
+    }
 
-      case .client(_):
-        throw RPCError(code: .failedPrecondition, message: "This worker has a client setup.")
+    mutating func startedServer(
+      _ server: GRPCServer,
+      stats: ServerStats,
+      eventLoopGroup: MultiThreadedEventLoopGroup
+    ) -> OnStartedServer {
+      let action: OnStartedServer
 
+      switch self.role {
       case .none:
-        self.role = .server(serverState)
+        let state = State.Server(server: server, stats: stats, eventLoopGroup: eventLoopGroup)
+        self.role = .server(state)
+        action = .runServer
+      case .server:
+        let error = RPCError(code: .alreadyExists, message: "A server has already been set up.")
+        action = .invalidState(error)
+      case .client:
+        let error = RPCError(code: .failedPrecondition, message: "This worker has a client setup.")
+        action = .invalidState(error)
       }
+
+      return action
     }
 
-    mutating func setupClients(
-      benchmarkClients: [BenchmarkClient],
+    enum OnStartedClients {
+      case runClients
+      case invalidState(RPCError)
+    }
+
+    mutating func startedClients(
+      _ clients: [BenchmarkClient],
       stats: ClientStats,
       rpcStats: RPCStats
-    ) throws {
-      let clientState = State.ClientState(
-        clients: benchmarkClients,
-        stats: stats,
-        rpcStats: rpcStats
-      )
+    ) -> OnStartedClients {
+      let action: OnStartedClients
+
       switch self.role {
-      case .server(_):
-        throw RPCError(code: .alreadyExists, message: "This worker has a server setup.")
+      case .none:
+        let state = State.Client(clients: clients, stats: stats, rpcStats: rpcStats)
+        self.role = .client(state)
+        action = .runClients
+      case .server:
+        let error = RPCError(code: .alreadyExists, message: "This worker has a server setup.")
+        action = .invalidState(error)
+      case .client:
+        let error = RPCError(
+          code: .failedPrecondition,
+          message: "Clients have already been set up."
+        )
+        action = .invalidState(error)
+      }
 
-      case .client(_):
-        throw RPCError(code: .failedPrecondition, message: "Clients have already been set up.")
+      return action
+    }
 
+    mutating func serverShutdown() -> EventLoopGroup? {
+      switch self.role {
+      case .client:
+        preconditionFailure("Invalid state")
+      case .server(let state):
+        self.role = .none
+        return state.eventLoopGroup
       case .none:
-        self.role = .client(clientState)
+        return nil
       }
     }
 
-    mutating func updateRPCStats() throws {
+    func stopListening() -> GRPCServer? {
```

Comment:
Similar to the above, maybe we can have an action here instead of the optional server?